Only create OwnerRRefs when processing remote calls

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31163 Only create OwnerRRefs when processing remote calls**

The purpose is to unblock integration with TorchScript. Currently,
an OwnerRRef will be created by either a remote call or a to_here
call, whichever arrives first. However, when making RRef an IValue,
we need to know the type of value held by the RRef, which is
retrived by checking the return type of the TorchScript function.
The TorchScript function is only avaible during the remote call
but not in the to_here() call. Hence, an OwnerRRef can only be
created when processing a remote call. This commit implements this
behavior by introducing a conditional variable for every OwnerRRef
in the RRefContext, and let the to_here() call and PyRRef::unpickle
block on the CV until the value is ready.

Differential Revision: [D18949591](https://our.internmc.facebook.com/intern/diff/D18949591)

mrshenli committed:
Only create OwnerRRefs when processing remote calls

The purpose is to unblock integration with TorchScript. Currently,
an OwnerRRef will be created by either a remote call or a to_here
call, whichever arrives first. However, when making RRef an IValue,
we need to know the type of value held by the RRef, which is
retrived by checking the return type of the TorchScript function.
The TorchScript function is only avaible during the remote call
but not in the to_here() call. Hence, an OwnerRRef can only be
created when processing a remote call. This commit implements this
behavior by introducing a conditional variable for every OwnerRRef
in the RRefContext, and let the to_here() call and PyRRef::unpickle
block on the CV until the value is ready.

[ghstack-poisoned]

mrshenli committed:
Update on "Only create OwnerRRefs when processing remote calls"


The purpose is to unblock integration with TorchScript. Currently,
an OwnerRRef will be created by either a remote call or a to_here
call, whichever arrives first. However, when making RRef an IValue,
we need to know the type of value held by the RRef, which is
retrived by checking the return type of the TorchScript function.
The TorchScript function is only avaible during the remote call
but not in the to_here() call. Hence, an OwnerRRef can only be
created when processing a remote call. This commit implements this
behavior by introducing a conditional variable for every OwnerRRef
in the RRefContext, and let the to_here() call and PyRRef::unpickle
block on the CV until the value is ready.

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit aca4a15:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=aca4a15b17fa3f08af2bc2ca7771a2c5d5ad3ce6).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3916041) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72791788))
```
HEAD is now at aca4a15b17 Only create OwnerRRefs when processing remote calls
+ git reset --hard aca4a15b17fa3f08af2bc2ca7771a2c5d5ad3ce6
HEAD is now at aca4a15b17 Only create OwnerRRefs when processing remote calls
+ git merge --allow-unrelated-histories --no-edit --no-ff d81c6bde3b32454bde336146ce0314f2a70ba076
Auto-merging torch/csrc/distributed/rpc/request_callback_impl.cpp
CONFLICT (content): Merge conflict in torch/csrc/distributed/rpc/request_callback_impl.cpp
Auto-merging test/rpc_test.py
Removing aten/src/THNN/generic/ELU.c
Removing aten/src/THCUNN/generic/ELU.cu
Removing aten/src/THCUNN/ELU.cu
Automatic merge failed; fix conflicts and then commit the result.
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

mrshenli committed:
Update on "Only create OwnerRRefs when processing remote calls"


The purpose is to unblock integration with TorchScript. Currently,
an OwnerRRef will be created by either a remote call or a to_here
call, whichever arrives first. However, when making RRef an IValue,
we need to know the type of value held by the RRef, which is
retrived by checking the return type of the TorchScript function.
The TorchScript function is only avaible during the remote call
but not in the to_here() call. Hence, an OwnerRRef can only be
created when processing a remote call. This commit implements this
behavior by introducing a conditional variable for every OwnerRRef
in the RRefContext, and let the to_here() call and PyRRef::unpickle
block on the CV until the value is ready.

Differential Revision: [D18949591](https://our.internmc.facebook.com/intern/diff/D18949591)

[ghstack-poisoned]

mrshenli committed:
Update on "Only create OwnerRRefs when processing remote calls"


The purpose is to unblock integration with TorchScript. Currently,
an OwnerRRef will be created by either a remote call or a to_here
call, whichever arrives first. However, when making RRef an IValue,
we need to know the type of value held by the RRef, which is
retrived by checking the return type of the TorchScript function.
The TorchScript function is only avaible during the remote call
but not in the to_here() call. Hence, an OwnerRRef can only be
created when processing a remote call. This commit implements this
behavior by introducing a conditional variable for every OwnerRRef
in the RRefContext, and let the to_here() call and PyRRef::unpickle
block on the CV until the value is ready.

Differential Revision: [D18949591](https://our.internmc.facebook.com/intern/diff/D18949591)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -138,6 +141,20 @@ class RRefContext {
+  // create OwnerRRef when processing remote calls.
+  // TODO: As OwnerRRef::getValue() is always called after
+  // OwnerRRef::setValue(), we should be able to remove the CV from OwnerRRef.
+  std::unordered_map<RRefId, std::condition_variable, RRefId::Hash> ownerCVs_;
```
Just to ask: would it be a lesser evil to just use a single condvar, shared across everybody in getOwnerRRef waiting on an rref? Realistically, how many are going to be waiting at once in condvar::wait?

The cost would be some spurious wakeups, but upon waking up it's just a hash table loop, and back to sleep. The extra work ends having an N^2 factor with respect to num waiting, so nothing we want to do if there are thousands waiting. But it might be better than either leaking memory forever, or the complexity of managing memory lifetime here?

Alternatively, if that approach is too simple, another approach would be to keep a  std::vector<std::condition_variable> of a small fixed size (say 16 or whatever), and share among rrefs by mod sharding (e.g. rrefId % num_vector_condvars)

mrshenli replied:
Thanks for commenting!

> Realistically, how many are going to be waiting at once in condvar::wait?

Whenever there are OwnerRRefs used before creation would have one waiting for this CV. This probably won't be too many and should be a transient stage for most cases. I agree this might be an overkill to keep this CV live for the entire lifetime of an OwnerRRef. Let me make changes. Thanks!

jjlilley replied:
Great, looks somewhat simpler now!

mrshenli committed:
Update on "Only create OwnerRRefs when processing remote calls"


The purpose is to unblock integration with TorchScript. Currently,
an OwnerRRef will be created by either a remote call or a to_here
call, whichever arrives first. However, when making RRef an IValue,
we need to know the type of value held by the RRef, which is
retrived by checking the return type of the TorchScript function.
The TorchScript function is only avaible during the remote call
but not in the to_here() call. Hence, an OwnerRRef can only be
created when processing a remote call. This commit implements this
behavior by introducing a conditional variable for every OwnerRRef
in the RRefContext, and let the to_here() call and PyRRef::unpickle
block on the CV until the value is ready.

Differential Revision: [D18949591](https://our.internmc.facebook.com/intern/diff/D18949591)

[ghstack-poisoned]

jjlilley commented:
Checking: are the failures in Phabricator related, or are they just flakes?

mrshenli commented:
@jjlilley I think they are irrelevant. None of the failed tests are using RPC I think.

wanchaol commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -202,6 +202,26 @@ template std::shared_ptr<OwnerRRef<IValue>> RRefContext::createOwnerRRef<
+    return std::static_pointer_cast<OwnerRRef<T>>(owners_[rrefId]);
+  } else {
+    // Scenario (2) retrieving an existing RRef
+    return std::static_pointer_cast<OwnerRRef<T>>(iter->second);
```
nit, we can get rid of the else branch and just return at the end.

jjlilley replied:
We could get rid of the else branch, though the existing code is probably marginally more efficient, since it avoids an extra table lookup.

wanchaol replied:
sounds good, let's keep it :)

wanchaol reviewed (approved):
lgtm. Thanks!

mrshenli commented:
Thanks for reviewing! Landing now.
